### PR TITLE
Make casesplit a HaRe action

### PR DESCRIPTION
**Pull request by Avi-D-coder**
Move `splitCaseCmd` from the GhcMod plugin to the HaRe plugin in order to avoid a circular dependency. 
Move the `casesplit`  tests to HaRe tests file. 
Add a `casesplit` CodeAction test.
I have created an associated [pull request](https://github.com/alanz/vscode-hie-server/pull/121) to avoid breaking vscode-hie-server use of `ghcmod:casesplit`.

closes: https://github.com/haskell/haskell-ide-engine/issues/907

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
What do you think about refactoring all the mkCodeActions in HaRe.hs?
see: Avi-D-coder/haskell-ide-engine@bd01ab3021789847f5022937f017c2fb689aeb75

**Comment by alanz** [comment]
I think the intention behind this ticket is to make the case split command generally available, by adding a case to the HaRe Plugin code action generator to offer a command to split the case, which activates this command.

So moving the mechanics of it to the HaRe Plugin is half the job, the other half is to modify [codeActionProvider](https://github.com/haskell/haskell-ide-engine/blob/master/src/Haskell/Ide/Engine/Plugin/HaRe.hs#L294) to generate the command.

Doing that would mean it is available in all HIE clients, and can be *removed* from the vscode client, as a special case.

File: src/Haskell/Ide/Engine/Plugin/HaRe.hs [hunk 1]
```diff
@@ -339,3 +410,9 @@ codeActionProvider pId docId _ _ (J.Range pos _) _ =
           title = "Duplicate definition of " <> name
       dupCmd <- mkLspCommand pId "dupdef" title (Just args)
       return $ J.CodeAction title (Just J.CodeActionRefactor) mempty Nothing (Just dupCmd)
+
+    mkCaseSplitAction name = do
+      let args = [J.toJSON $ HP (docId ^. J.uri) pos]
+          title = "Case split on " <> name
+      splCmd <- mkLspCommand pId "casesplit" title (Just args)
+      return $ J.CodeAction title (Just J.CodeActionRefactorRewrite) mempty Nothing (Just splCmd)
```
Comment (Avi-D-coder) [inline src/Haskell/Ide/Engine/Plugin/HaRe.hs]: @alanz This generates the command. It works on neovim.

File: src/Haskell/Ide/Engine/Plugin/HaRe.hs [hunk 2]
```diff
@@ -339,3 +410,9 @@ codeActionProvider pId docId _ _ (J.Range pos _) _ =
           title = "Duplicate definition of " <> name
       dupCmd <- mkLspCommand pId "dupdef" title (Just args)
       return $ J.CodeAction title (Just J.CodeActionRefactor) mempty Nothing (Just dupCmd)
+
+    mkCaseSplitAction name = do
```
Comment (Avi-D-coder) [inline src/Haskell/Ide/Engine/Plugin/HaRe.hs]: @alanz This generates the command. It works on neovim.

File: src/Haskell/Ide/Engine/Plugin/HaRe.hs [hunk 3]
```diff
@@ -339,3 +410,9 @@ codeActionProvider pId docId _ _ (J.Range pos _) _ =
...
+      let args = [J.toJSON $ HP (docId ^. J.uri) pos]
+          title = "Case split on " <> name
```
Comment (Avi-D-coder) [inline src/Haskell/Ide/Engine/Plugin/HaRe.hs]: @alanz This generates the command. It works on neovim.

**Comment by alanz** [comment]
Oops, sorry. I saw the matching change on vscode, and assumed it was just a switch over.

I think to ease the transition we should keep it exposed in ghc-mod for the time being.  This would involve re-instating the command there, and getting it to call the function from HaRe. Which hopefully does not cause an import loop.

**Comment by Avi-D-coder** [comment]
I'll give moving splitCaseCmd and HarePoint to a new module a try tomorrow. I think that will avoid circular dependencies. 

It's easy to miss in all the duplicate mkCodeAction definitions. That's why I'm advocating for the linked refactor.

**Comment by alanz** [comment]
The problem with the linked refactor is that not everyone updates their hie installation all that often, and vscode pushed updates pretty much immediately.

So we need to support the old and the new for a while, else things will break.

**Comment by alanz** [comment]
In fact, the simplest might be to leave the command in the ghc-mod plugin, but just add the code action generation to it.

Then we can make an issue to migrate it to the HaRe module at a future date, when we have removed the explicit code from the vscode plugin.

**Comment by lorenzo** [comment]
Can we have the ghc-mod action trigger the now migrated HaRe action via the client? I think that would be the cleanest migration path

File: src/Haskell/Ide/Engine/Plugin/HaRe.hs [hunk 4]
```diff
@@ -294,20 +359,26 @@ hoist f a =
...
             , mkDuplicateAction name
             ]
-        else return (IdeResultOk [])
+        _   -> case getArtifactsAtPos pos (locMap info) of
+               [h] -> do
+                let name = Hie.showName $ snd h
+                debugm $ show name
+                IdeResultOk <$> sequence [
+                  mkCaseSplitAction name
+                  ]
```
Comment (bubba) [inline src/Haskell/Ide/Engine/Plugin/HaRe.hs]: The command generation stuff LGTM

**Comment by Avi-D-coder** [comment]
@lorenzo  I'm not clear how we could call the HaRa command with a GhcMod command. What I have done is move anything that would case circular dependence to HieExtras.

**Comment by Avi-D-coder** [comment]
@alanz I rebased the refactor on the new changes that don't break compatibility. What are your thoughts on: Avi-D-coder/haskell-ide-engine@9f910841646a75f9ef8bf8cbf66b274be2159b20

**Comment by alanz** [comment]
@Avi-D-coder Its looking good, and the refactoring of the HaRe command construction makes it a lot clearer, in my opinion.
I suggest you either add it to this PR or make a new PR with it.

**Comment by lorenzo** [comment]
@Avi-D-coder I was meaning something like this https://github.com/haskell/haskell-ide-engine/pull/889/files#diff-df7cc8b4c1792382d5165b401f9ba6d0R536

**Comment by alanz** [comment]
So, playing with it, I see we test that the case split command happens, but not that a code action is generated.

It should probably be tested together with the other code actions in https://github.com/haskell/haskell-ide-engine/blob/master/test/functional/FunctionalCodeActionsSpec.hs#L226

**Pull request merged**